src/main/java/org/dependencytrack/tasks/SendScheduledNotificationTask.java

```diff
@@ -0,0 +1,224 @@
+                        }
+                        atLeastOneSuccessfulPublish |= true;
+                    } else {
+                        LOGGER.error("The defined notification publisher is not assignable from " + Publisher.class.getCanonicalName() + " (%s)".formatted(ruleCtx));
+                        errorsDuringExecution |= true;
+                    }
+                } catch (ClassNotFoundException | NoSuchMethodException | InstantiationException
+                        | InvocationTargetException | IllegalAccessException e) {
+                    LOGGER.error("An error occurred while instantiating a notification publisher (%s)".formatted(ruleCtx), e);
+                    errorsDuringExecution |= true;
+                } catch (PublisherException publisherException) {
+                    LOGGER.error("An error occurred during the publication of the notification (%s)".formatted(ruleCtx), publisherException);
+                    errorsDuringExecution |= true;
+                }
+            }
+            if (!errorsDuringExecution || atLeastOneSuccessfulPublish) {
+                /*
+                 * Update last execution time after successful operation (even without
+                 * publishing) to avoid duplicate notifications in the next run and signalize
+                 * user indirectly, that operation has ended without failure
+                 */
+                qm.updateScheduledNotificationRuleLastExecutionTimeToNowUtc(rule);
+                LOGGER.info("Successfuly processed notification publishing for scheduled notification rule " + scheduledNotificationRuleUuid);
+            }
+            else {
+                LOGGER.error("Errors occured while processing notification publishing for scheduled notification rule " + scheduledNotificationRuleUuid);
+            }
+        }
+        catch (Exception e) {
+            LOGGER.error("An error occurred while processing scheduled notification rule " + scheduledNotificationRuleUuid, e);
+        }
+    }
+
+    private List<Project> evaluateAffectedProjects(QueryManager qm, ScheduledNotificationRule rule) {
+        List<Project> affectedProjects;
+        /* 
+         * TODO:
+         * To workaround the inconsitent parent-child relationship in projects delivered
+         * by QueryManager.getAllProjects() (and some other multi-project methods), we
+         * need to retrieve them one by one by their UUIDs. This way it was empirically
+         * proven that the parent-child relationship is (more) consistent.
+         */
+        if(rule.getProjects().isEmpty()){
+            // if rule does not limit to specific projects, get all projects and their children, if configured
+            affectedProjects = qm.detach(qm.getAllProjects())
+                    .stream()
+                    .filter(p -> rule.isNotifyChildren() ? true : p.getParent() == null)
+                    .collect(Collectors.toList());
```

Comment:
We might avoid that, if you have a good idea how. What we really need in the end is the project `id`, `uuid` and the basic information for `toJson(Project)` for the template. In the basic mail template, we need the project `name` and `version` (if i didn't forget anything). The last thing we need to have, is the proper children-connection to find all projects with their children, if the setting in the scheduled notification rule is active.

A different approach here might be the following: instead of getting new findings since date x for each of the affected projects, we could fetch all findings since date x and then filter them similar to `NotificationRouter.resolveRules(...)`:
if(rule has projects)
-> get only findings, which have a project id of the rule projects
else
-> use all findings to get the affected project ids, load them and use this project list for the scheduled notification.

This way, we only load what is needed. But what is important nonetheless: we need a proper children-connection in `rule.getProjects()`. If that's working fine, this approach should work well.

What do you think about that?

Reply:
This should be "fixed" by now, as i introduced the ScheduledNotificationFactory, which will only fetch the projects that are included in the findings and uses the data from this database call to extract all information for the sub-parts needed for the email-template. This way, the performance impact should be small.